open_academy/models/open_academy_session.py

```diff
@@ -1,13 +1,81 @@
-from odoo import models, fields
+from odoo import models, fields, api, exceptions
+from datetime import timedelta
 
 
 class Session(models.Model):
     _name = "openacademy.session"
     _description = "Session"
 
-    name = fields.Char(string="Session", required=True)
-    start_date = fields.Date()
-    duration = fields.Float()
+    active = fields.Boolean(default=True)
+    attendee_ids = fields.Many2many('res.partner')
+    course_id = fields.Many2one('openacademy.course')    
+    duration = fields.Integer()
+    end_date = fields.Date(
+        store=True,
+        compute='_compute_end_date',
+        inverse='_inverse_end_date'
+    )
+    instructor_id = fields.Many2one(
+        'res.partner',
+        ondelete="set null",
+        domain=[
+            '|',
+            ('instructor', '=', True),
+            ('category_id.name', 'ilike', "Teacher")
+        ]
+    )
+    name = fields.Char(string="Session", required=True)    
     seats = fields.Integer(string="Number of seats")
-    instructor_id = fields.Many2one('res.partner', ondelete="set null")
-    course_id = fields.Many2one('openacademy.course', ondelete='')
+    start_date = fields.Date(default=fields.Date.today)
+    taken_seats = fields.Float(compute='_compute_taken_seats')
+    
+    @api.depends('start_date', 'duration')
+    def _compute_end_date(self):
+        for record in self:
+            if not (record.start_date and record.duration):
+                record.end_date = record.start_date
+                continue
+            start = fields.Datetime.from_string(record.start_date)
+            duration = timedelta(days=record.duration, seconds=-1)
+            record.end_date = start + duration
+
+    def _inverse_end_date(self):
+        for record in self:
+            if not (record.start_date and record.end_date):
+                continue
+            
+            start_date = fields.Datetime.from_string(record.start_date)
+            end_date = fields.Datetime.from_string(record.end_date)
+            record.duration = (end_date - start_date).days + 1
+
+    @api.depends('seats', 'attendee_ids')
+    def _compute_taken_seats(self):
+        for record in self:
+            if not record.seats:
+                record.taken_seats = 0.0
+            else:
+                record.taken_seats = 100.0 * len(record.attendee_ids) / record.seats
+
+    @api.onchange('seats', 'attendee_ids')
+    def _onchange_valid_seats(self):
```

Comment:
Guidelines says: Onchange method: the onchange method pattern is `_onchange_<field_name>`

This should be:
```suggestion
    def _onchange_seats(self):
```